source/iterator/snapshot.go

```diff
@@ -0,0 +1,168 @@
+// Copyright © 2022 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package iterator
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	sdk "github.com/conduitio/conduit-connector-sdk"
+
+	"github.com/conduitio-labs/conduit-connector-firebolt/client"
+	"github.com/conduitio-labs/conduit-connector-firebolt/source/position"
+)
+
+const (
+	// metadata related.
+	metadataTable = "firebolt.table"
+)
+
+// SnapshotIterator snapshot iterator.
+type SnapshotIterator struct {
+	// firebolt client.
+	client *client.Client
+	// index - current index of element in current batch which iterator converts to record.
+	indexInBatch int
+	// batchID - current batch id, show what batch iterator uses, using in query to get currentBatch.
+	batchID int
+	// batchSize size of batch.
+	batchSize int
+	// currentBatch - rows in current batch from table.
+	currentBatch []map[string]any
+	// name of column what iterator use for setting key in record.
+	primaryKey string
+	// list of columns to reading from table.
+	columns []string
+	// table which iterator read.
+	table string
+}
+
+func NewSnapshotIterator(
+	client *client.Client,
+	batchSize int,
+	columns []string,
+	table, primaryKey string,
+) *SnapshotIterator {
+	return &SnapshotIterator{
+		client:     client,
+		batchSize:  batchSize,
+		primaryKey: primaryKey,
+		columns:    columns,
+		table:      table}
+}
+
+// Setup iterator.
+func (i *SnapshotIterator) Setup(ctx context.Context, p sdk.Position) error {
+	var index, batchID int
+
+	if p != nil {
+		pos, err := position.ParseSDKPosition(p)
+		if err != nil {
+			return nil
+		}
+
+		index = pos.IndexInBatch + 1
+		batchID = pos.BatchID
+	}
+
+	i.indexInBatch = index
+	i.batchID = batchID
+
+	rows, err := i.client.GetRows(ctx, i.table, i.primaryKey, i.columns, i.batchSize, i.batchID)
```

Comment:
added

Reply:
Thanks! Maybe I'm nitpicky, but I'd like to suggest one more thing, and that is to use Str() instead of Msgf(). The logs are structured, and if we add a separate key for the parameters, it makes it easier to search for a table for example. But I like Msgf() too.:)

Reply:
Changed to use Str()